### PR TITLE
Classical ml_Draft

### PR DESCRIPTION
**Pull request by KelllyChen**
### Description of each script
- extract_features.py: embed all images and save them in image_features.csv
- cluster_items.py: use kmeans to cluster all images into three categories(ideally top, bottom, others), output as image_clusters.csv
- llava_random.ipynb: make all _1 images as input and complete the outfit through randomly pick an item in the two other clusters, then use llava to score these outfit from 1-10, save the matching and score in llava_outfit_scores.csv
- ml_model.py: take llava_outfit_scores.csv and split them into train, validation, test sets and train a RandomForestRegressor model based on these outfit(X) and the score they got from llava(y), and keep test set untouched, save all test images into test_anchor_images.csv
- llava_ml_model.ipynb: make those image in test_anchor_images.csv as input, and this time complete the outfit based on coisine similarity(but still pick one in each cluster), keep the top 3 pairs, and use the saved RandomForestRegressor model to grade them and pick the best one. Then feed these outfit to llava again to see the score

### Issues
- In llava_random.ipynb, I tried different prompt to let llava give a score from 1-10(for example give instructions on what kind of outfit should fall in what score ranges), but it still only gave me score from 5-7
- The final average score i got from llava in llava_ml_model.ipynb is 10, which apparantely has some problem in it. Due to the limitation of kmeans, some items were put into the wrong categories(like top become bottom) tho this is not the main issue I want to solve here. Since this limitation, some complete outfit would contain 3 tops and it should be given a low score, however llava gave a score of 10

### Ideas to solve the issuse
- Basically i think the way to solve the issues is letting llava give a more valid score, and as my understaing, it would lie on a better prompt. If llava can grade the outfit better, the training data would be better and the the same prompt used for final grading should make more sense.

### CODE REVIEW

**Comment by choonghwanlee** [comment]
like the idea of using Llava to generate pre-trained image embeddings as features to your ML approach and also as a LLM-as-a-judge at the end to grade outfits. but aren't categories for each product given in the dataset? and for each product, two complementary products are already given? not sure why Llava is needed to generate the outfits for training your Random Forest model. can't we just use the information in our dataset to identify pairs/outfits of complementary and non-complementary products?

**Comment by KelllyChen** [comment]
1. In data.xlsx, only main items have categories. Although their matching items have name, but the way the website named these items is too diverse and not easy to categorize or match to categories of main items, so I decided to use kmeans and put all the images in three cluster(top, bottom, others). Quality might not be that good but an easier way to group the images.

2. Though there are matching items in data.xlsx, like we discussed earlier, there's a limitation in these matchings. I define a complete outfit as top+bottom+others(might be shoes, sunglasses, perfume, it's vague but basically any items in the cluster "others" ). However, in data.xlsx, the matching it gave sometimes is top+two bottoms or top+perfume, and this doesn't form a complete outfit. As a result, I need sveral complete outfits that contain one item in each cluster to be used as ml model's training data

**Comment by choonghwanlee** [comment]
For 1, I just created a file that goes through the data.xlsx and identifies the categories of each item, so even recommended items have categories associated with them. perhaps we can use that?

For 2. totally agree with the limitations. I think we can simplify the problem by training a classifier to determine if any two products are complementary. and during inference time, we can find the top pants and top shoes for a t-shirt the user inputs for example. so instead of learning to classify an entire output, it compares two items for compatibility, and we can repeat this process for each category we want.

**Comment by KelllyChen** [comment]
1. Could you push the file first so I can take a look of it
2. yes but we still need to let llava score the entire complete outfit, is that right? so the method would be similar to what I have now, only the ml model treats t shirt+pants and t shirt+socks separately, but not t shirt+pants+socks

And also still think the issues on my pipeline now more lie on how llava gave score, so i might still first focus on this part to see how i can let llava give a score that makes more sense. Once i can get this done, i think the overall process on the classical ml recommendation should be fine

**Comment by KelllyChen** [comment]
I adjust the prompt and and now the scores llava gave make more sense, so I keep the same pipeline for classical ml approach. Another thing is that since it took a long time for llava to go through all images, so I randomly sample 700 main images from the original 3520 main images as training data to reduce the time needed for generating score for training RandomForestRegressor

**Comment by ludaladila** [comment]
Adjusting the prompt to get more reasonable LLaVA scores makes a lot of sense, and keeping the pipeline consistent sounds like a solid approach. Nice work!
